processor/processor_geolocation_test.go

```diff
@@ -0,0 +1,305 @@
+
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/require"
+	"github.com/tidwall/gjson"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	kithttputil "github.com/rudderlabs/rudder-go-kit/httputil"
+	kithelper "github.com/rudderlabs/rudder-go-kit/testhelper"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/rand"
+	"github.com/rudderlabs/rudder-server/runner"
+	"github.com/rudderlabs/rudder-server/testhelper/backendconfigtest"
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+	"github.com/rudderlabs/rudder-server/testhelper/transformertest"
+)
+
+func TestProcessorGeolocation(t *testing.T) {
+	const (
+		boxfordIP = "2.125.160.216"
+		londonIP  = "81.2.69.142"
+		unknownIP = "22.125.160.216"
+		invalidIP = "invalid"
+	)
+
+	t.Run("feature disabled", func(t *testing.T) {
+		new(geolocationScenario).
+			WithGeolocationFeatureEnabled(false).
+			WithGeolocationEnabledAtSource(true).
+			WithClientIP(boxfordIP).
+			WithContextIP(londonIP).
+			Run(t, func(t *testing.T, event json.RawMessage) {
+				require.Empty(t, gjson.GetBytes(event, "context.geo").Raw, "no geolocation information should be present when the feature is disabled")
+			})
+	})
+
+	t.Run("feature enabled but geolocation disabled at source", func(t *testing.T) {
+		new(geolocationScenario).
+			WithGeolocationFeatureEnabled(true).
+			WithGeolocationEnabledAtSource(false).
+			WithClientIP(boxfordIP).
+			WithContextIP(londonIP).
+			Run(t, func(t *testing.T, event json.RawMessage) {
+				require.Empty(t, gjson.GetBytes(event, "context.geo").Raw, "no geolocation information should be present when geolocation is disabled at source")
+			})
+	})
+
+	t.Run("enabled with known clientIP and context.ip", func(t *testing.T) {
+		new(geolocationScenario).
+			WithGeolocationFeatureEnabled(true).
+			WithGeolocationEnabledAtSource(true).
+			WithClientIP(boxfordIP).
+			WithContextIP(londonIP).
+			Run(t, func(t *testing.T, event json.RawMessage) {
+				require.NotEmpty(t, gjson.GetBytes(event, "context.geo").Raw, string(event), "geolocation information should be present")
+				require.Equal(t, "London", gjson.GetBytes(event, "context.geo.city").String(), "contex.ip should take precedence over clientIP")
+			})
+	})
+
+	t.Run("enabled with known clientIP without context.ip", func(t *testing.T) {
+		new(geolocationScenario).
+			WithGeolocationFeatureEnabled(true).
+			WithGeolocationEnabledAtSource(true).
+			WithClientIP(boxfordIP).
+			WithContextIP("").
+			Run(t, func(t *testing.T, event json.RawMessage) {
+				require.NotEmpty(t, gjson.GetBytes(event, "context.geo").Raw, string(event), "geolocation information should be present")
+				require.Equal(t, "Boxford", gjson.GetBytes(event, "context.geo.city").String(), "clientIP should be used by the geolocation service")
+			})
+	})
+
+	t.Run("enabled with invalid context.ip but valid clientIP", func(t *testing.T) {
```

Comment:
Not sure if this should be the desired behaviour, given that transformer doesn't perform any kind of validation on these fields. So maybe we should just have a precedence order and don't ignore invalid ips

Reply:
Yeah, that's what I am thinking, we shouldn't check if the values are valid or not, just pass the first available value to locate function.